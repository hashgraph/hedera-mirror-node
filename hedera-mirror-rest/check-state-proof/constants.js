/*-
 *
 * Hedera Mirror Node
 *  ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 *  ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

const MAX_RECORD_LENGTH = 64 * 1024;

const MAX_TRANSACTION_LENGTH = 64 * 1024;

const SHA_384_LENGTH = 48;

const SHA_384_WITH_RSA = {
  type: 1,
  maxLength: 384,
};

const SIMPLE_SUM = 101;

module.exports = {
  MAX_RECORD_LENGTH,
  MAX_TRANSACTION_LENGTH,
  SHA_384_LENGTH,
  SHA_384_WITH_RSA,
  SIMPLE_SUM,
};
