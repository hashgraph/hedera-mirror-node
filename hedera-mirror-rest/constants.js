/*-
 * ‌
 * Hedera Mirror Node
 *​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 *​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

'use strict';

// url query filer keys
const filterKeys = {
  ACCOUNT_ID: 'account.id',
  ACCOUNT_BALANCE: 'account.balance',
  ACCOUNT_PUBLICKEY: 'account.publickey',
  ENCODING: 'encoding',
  ENTITY_PUBLICKEY: 'publickey',
  LIMIT: 'limit',
  ORDER: 'order',
  RESULT: 'result',
  SEQUENCE_NUMBER: 'sequencenumber',
  TIMESTAMP: 'timestamp',
  TOKENID: 'tokenid',
  TOKEN_ID: 'token.id',
  CREDIT_TYPE: 'type',
  TRANSACTION_TYPE: 'transactiontype',
};

// sql table columns
const entityColumns = {
  ENTITY_NUM: 'entity_num',
  ENTITY_REALM: 'entity_realm',
  ENTITY_SHARD: 'entity_shard',
  PUBLIC_KEY: 'ed25519_public_key_hex',
};

const transactionColumns = {
  TYPE: 'type',
};

const requestIdLabel = 'requestId';
const responseDataLabel = 'mirrorRestData';

const orderFilterValues = {
  ASC: 'asc',
  DESC: 'desc',
};

// topic messages filter options
const characterEncoding = {
  BASE64: 'base64',
  UTF8: 'utf-8',
};

const transactionResultFilter = {
  SUCCESS: 'success',
  FAIL: 'fail',
};

const cryptoTransferType = {
  CREDIT: 'credit',
  DEBIT: 'debit',
};

const cloudProviders = {
  S3: 'S3',
  GCP: 'GCP',
};

const defaultCloudProviderEndpoints = {
  [cloudProviders.S3]: 'https://s3.amazonaws.com',
  [cloudProviders.GCP]: 'https://storage.googleapis.com',
};

const networks = {
  DEMO: 'DEMO',
  MAINNET: 'MAINNET',
  TESTNET: 'TESTNET',
  PREVIEWNET: 'PREVIEWNET',
  OTHER: 'OTHER',
};

const defaultBucketNames = {
  [networks.DEMO]: 'hedera-demo-streams',
  [networks.MAINNET]: 'hedera-mainnet-streams',
  [networks.TESTNET]: 'hedera-stable-testnet-streams-2020-08-27',
  [networks.PREVIEWNET]: 'hedera-preview-testnet-streams',
  [networks.OTHER]: null,
};

const recordStreamPrefix = 'recordstreams/record';

module.exports = {
  characterEncoding,
  cloudProviders,
  cryptoTransferType,
  defaultBucketNames,
  defaultCloudProviderEndpoints,
  entityColumns,
  filterKeys,
  networks,
  orderFilterValues,
  recordStreamPrefix,
  requestIdLabel,
  responseDataLabel,
  transactionColumns,
  transactionResultFilter,
};
