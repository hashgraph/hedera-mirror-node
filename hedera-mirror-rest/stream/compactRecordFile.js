/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import crypto from 'crypto';
import _ from 'lodash';

import HashObject from './hashObject';
import RecordFile from './recordFile';
import RecordStreamObject from './recordStreamObject';
import {calculateRunningHash} from './runningHash';

const {SHA_384} = HashObject;

class CompactRecordFile extends RecordFile {
  static compactObjectFields = [
    'head',
    'startRunningHashObject',
    'hashesBefore',
    'recordStreamObject',
    'hashesAfter',
    'endRunningHashObject',
    'blockNumber',
  ];

  constructor(bufferOrObj) {
    super();

    if (!this.constructor._support(bufferOrObj)) {
      throw new Error(`Unsupported record file version, expect ${this.constructor.version}`);
    }

    if (Buffer.isBuffer(bufferOrObj)) {
      this._parseFromBuffer(bufferOrObj);
    } else {
      this._parseFromObj(bufferOrObj);
    }
  }

  static _support(bufferOrObj) {
    const buffer = Buffer.isBuffer(bufferOrObj) ? bufferOrObj : bufferOrObj.head;
    return RecordFile._readVersion(buffer) === this.version;
  }

  static canCompact(buffer) {
    return Buffer.isBuffer(buffer) && this._support(buffer);
  }

  toCompactObject(transactionId, nonce = 0, scheduled = false) {
    if (!this.containsTransaction(transactionId, nonce, scheduled)) {
      throw new Error(`Transaction ${transactionId} not found in the successful transactions map`);
    }

    if (this._hasRecordStreamObjects()) {
      // parsed from a full record file
      const transactionKey = RecordFile._getTransactionKey(transactionId, nonce, scheduled);
      const index = this._transactionMap[transactionKey];
      this.recordStreamObject = this._getRecordStreamObject(index);

      this._hashes.forEach((value, current) => {
        if (!value && current !== index) {
          // calculate and cache the hash if not found and this is not the transaction of interest
          const recordStreamObject = this._getRecordStreamObject(current);
          this._hashes[current] = crypto.createHash(SHA_384.name).update(recordStreamObject).digest();
        }
      });
      this.hashesBefore = this._hashes.slice(0, index);
      this.hashesAfter = this._hashes.slice(index + 1);
    }

    return _.pick(this, this.constructor.compactObjectFields);
  }

  /**
   * Calculates the record file metadata hash
   *
   * @returns {Buffer}
   */
  _calculateMetadataHash() {
    throw new Error('Unsupported operation');
  }

  /**
   * Gets the RecordStreamObject at the index
   *
   * @param {Number} index
   * @returns {Buffer}
   */
  _getRecordStreamObject(index) {
    throw new Error('Unsupported operation');
  }

  /**
   * Whether the record file object has RecordStreamObjects
   *
   * @return {boolean}
   */
  _hasRecordStreamObjects() {
    return false;
  }

  /**
   * Parses the record file from the raw buffer
   *
   * @param {Buffer} buffer
   * @throws {Error}
   */
  _parseFromBuffer(buffer) {
    throw new Error('Unsupported operation');
  }

  /**
   * Parses the record file from the compact format object
   *
   * @param obj
   */
  _parseFromObj(obj) {
    Object.assign(this, _.pick(obj, this.constructor.compactObjectFields));

    this._verifyEndRunningHash();

    // add the transaction to the map
    const recordStreamObject = new RecordStreamObject(this.recordStreamObject);
    this._addTransaction(recordStreamObject.record);

    // calculate metadata hash
    this._metadataHash = this._calculateMetadataHash();
  }

  /**
   * Verifies the calculated end running hash matches what's in the record file
   *
   * @throws {Error} Will throw if calculated hash doesn't match
   */
  _verifyEndRunningHash() {
    const startHashObject = new HashObject(this.startRunningHashObject);
    const endHashObject = new HashObject(this.endRunningHashObject);
    const hashes = [
      startHashObject.hash,
      ...this.hashesBefore,
      crypto.createHash(SHA_384.name).update(this.recordStreamObject).digest(),
      ...this.hashesAfter,
    ];

    // when calculating running hash, classId and classVersion are digested in little-endian
    const header = startHashObject.getHeaderLE();
    const actualHash = hashes.reduce((runningHash, nextHash) =>
      calculateRunningHash(
        {
          header,
          hash: runningHash,
        },
        {
          header,
          hash: nextHash,
        },
        SHA_384.name
      )
    );
    if (!actualHash.equals(endHashObject.hash)) {
      throw new Error('End object running hash mismatch');
    }
  }
}

export default CompactRecordFile;
