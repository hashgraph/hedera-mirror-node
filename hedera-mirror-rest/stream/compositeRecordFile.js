/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import RecordFile from './recordFile';
import RecordFilePreV5 from './recordFilePreV5';
import RecordFileV5 from './recordFileV5';
import RecordFileV6 from './recordFileV6';

const delegates = [RecordFilePreV5, RecordFileV5, RecordFileV6];

class CompositeRecordFile extends RecordFile {
  constructor(bufferOrObj) {
    super();

    const clazz = delegates.reduce((match, cls) => {
      return match || (cls._support(bufferOrObj) ? cls : null);
    }, null);
    if (!clazz) {
      throw new Error('Unsupported record file');
    }

    this.delegate = new clazz(bufferOrObj);
  }

  static canCompact(bufferOrObj) {
    return delegates.reduce((compactable, clazz) => {
      return compactable || clazz.canCompact(bufferOrObj);
    }, false);
  }

  containsTransaction(transactionId, nonce = 0, scheduled = false) {
    return this.delegate.containsTransaction(transactionId, nonce, scheduled);
  }

  getFileHash() {
    return this.delegate.getFileHash();
  }

  getMetadataHash() {
    return this.delegate.getMetadataHash();
  }

  getTransactionMap() {
    return this.delegate.getTransactionMap();
  }

  getVersion() {
    return this.delegate.getVersion();
  }

  toCompactObject(transactionId, nonce = 0, scheduled = false) {
    return this.delegate.toCompactObject(transactionId, nonce, scheduled);
  }
}

export default CompositeRecordFile;
