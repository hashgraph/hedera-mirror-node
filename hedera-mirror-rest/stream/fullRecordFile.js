/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2022 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

// external libraries
const crypto = require('crypto');

const {BYTE_SIZE, INT_SIZE} = require('./constants');
const HashObject = require('./hashObject');
const RecordFile = require('./recordFile');
const RecordStreamObject = require('./recordStreamObject');
const {readLengthAndBytes} = require('./utils');

const {MAX_TRANSACTION_LENGTH, MAX_RECORD_LENGTH} = RecordStreamObject;
const {SHA_384} = HashObject;

// version, hapiVersion, previous hash marker, SHA-384 hash
const PRE_V5_HEADER_LENGTH = INT_SIZE + INT_SIZE + BYTE_SIZE + SHA_384.length;

class FullRecordFile extends RecordFile {
  /**
   * Parses rcd file storing hash and transactionId map for future verification
   * @param {Buffer} buffer
   */
  constructor(buffer) {
    super();

    if (FullRecordFile._support(buffer)) {
      this._parsePreV5RecordFile(buffer);
    } else {
      throw new Error(`Unsupported record file`);
    }
  }

  static _support(bufferOrObj) {
    if (!Buffer.isBuffer(bufferOrObj)) {
      return false;
    }

    const version = RecordFile._readVersion(bufferOrObj);
    return version === 1 || version === 2;
  }

  _parsePreV5RecordFile(buffer) {
    this._version = RecordFile._readVersion(buffer);
    this._calculatePreV5FileHash(buffer.readInt32BE(), buffer);

    buffer = buffer.slice(PRE_V5_HEADER_LENGTH);
    let index = 0;
    while (buffer.length !== 0) {
      const marker = buffer.readInt8();
      if (marker !== 2) {
        throw new Error(`Unsupported marker ${marker}, expect 2`);
      }

      buffer = buffer.slice(BYTE_SIZE);
      const transaction = readLengthAndBytes(buffer, BYTE_SIZE, MAX_TRANSACTION_LENGTH, false);
      const record = readLengthAndBytes(buffer.slice(transaction.length), BYTE_SIZE, MAX_RECORD_LENGTH, false);
      this._addTransaction(record.bytes, index);
      index++;

      buffer = buffer.slice(transaction.length + record.length);
    }
  }

  _calculatePreV5FileHash(version, buffer) {
    const fileDigest = crypto.createHash(SHA_384.name);

    if (version === 1) {
      fileDigest.update(buffer);
    } else {
      // version 2
      const contentHash = crypto.createHash(SHA_384.name).update(buffer.slice(PRE_V5_HEADER_LENGTH)).digest();
      fileDigest.update(buffer.slice(0, PRE_V5_HEADER_LENGTH)).update(contentHash);
    }

    this._fileHash = fileDigest.digest();
  }
}

module.exports = FullRecordFile;
