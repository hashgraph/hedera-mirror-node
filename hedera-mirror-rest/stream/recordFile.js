/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {proto} from '@hashgraph/proto';
import {logger, protoTransactionIdToTransactionId} from './utils';

class RecordFile {
  static version;

  constructor() {
    this._fileHash = null;
    this._metadataHash = null;
    // a map of successful transactions, from its transaction ID to its index in the parsed transactions array
    // if the concrete class doesn't implement a transaction array, the index will be null
    this._transactionMap = {};
  }

  static _getTransactionKey(transactionId, nonce, scheduled) {
    return `${transactionId}-${nonce}-${scheduled}`;
  }

  static _readVersion(buffer) {
    return buffer.readInt32BE();
  }

  static _support(bufferOrObj) {
    return false;
  }

  static canCompact(buffer) {
    return false;
  }

  /**
   * Checks if a transaction is in the record file's successful transaction map
   *
   * @param {TransactionId} transactionId
   * @param {Number} nonce
   * @param {boolean} scheduled
   * @returns {boolean}
   */
  containsTransaction(transactionId, nonce = 0, scheduled = false) {
    return RecordFile._getTransactionKey(transactionId, nonce, scheduled) in this._transactionMap;
  }

  /**
   * Gets the record file's file hash, may return null if not available.
   *
   * @returns {Buffer}
   */
  getFileHash() {
    return this._fileHash;
  }

  /**
   * Gets the record file's metadata hash, may return null if not available.
   *
   * @returns {Buffer}
   */
  getMetadataHash() {
    return this._metadataHash;
  }

  /**
   * Gets the transaction map.
   * @returns {{}}
   */
  getTransactionMap() {
    return this._transactionMap;
  }

  /**
   * Gets the version.
   *
   * @returns {number}
   */
  getVersion() {
    return this.constructor.version;
  }

  /**
   * Converts the parsed record file to the compact format object if possible. Throws error if the transactionId is not
   * in the successful transaction map or the implementation does not support the operation.
   *
   * @param {TransactionId} transactionId
   * @param {Number} nonce
   * @param {boolean} scheduled
   * @returns {{}}
   */
  toCompactObject(transactionId, nonce = 0, scheduled = false) {
    throw new Error('Unsupported operation');
  }

  /**
   * Adds a mapping of a transaction's id to its index in the record file if the transaction response status is SUCCESS.
   *
   * @param {Buffer|proto.TransactionRecord} recordBufferOrObject
   * @param {Number} index
   * @private
   */
  _addTransaction(recordBufferOrObject, index = null) {
    const record =
      recordBufferOrObject instanceof Buffer
        ? proto.TransactionRecord.decode(recordBufferOrObject)
        : recordBufferOrObject;
    const {receipt, transactionID, scheduleRef} = record;
    const transactionId = protoTransactionIdToTransactionId(transactionID);
    const scheduled = scheduleRef !== null;
    if (
      receipt.status !== proto.ResponseCodeEnum.SUCCESS &&
      receipt.status !== proto.ResponseCodeEnum.FEE_SCHEDULE_FILE_PART_UPLOADED &&
      receipt.status !== proto.ResponseCodeEnum.SUCCESS_BUT_MISSING_EXPECTED_OPERATION
    ) {
      logger.debug(`Skip non-successful transaction ${transactionId}, ${receipt.status}`);
      return;
    }

    const transactionKey = RecordFile._getTransactionKey(transactionId, transactionID.nonce, scheduled);
    logger.debug(`Add successful transaction ${transactionKey}`);
    this._transactionMap[transactionKey] = index;
  }
}

export default RecordFile;
