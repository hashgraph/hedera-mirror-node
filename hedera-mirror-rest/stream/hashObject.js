/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {INT_SIZE} from './constants';
import StreamObject from './streamObject';
import {readLengthAndBytes} from './utils';

class HashObject extends StreamObject {
  // properties of SHA-384 hash algorithm
  static SHA_384 = {
    encoding: 'hex',
    length: 48,
    name: 'sha384',
  };

  /**
   * Reads the body of the hash object
   * @param {Buffer} buffer
   * @returns {Number} The size of the body in bytes
   */
  _readBody(buffer) {
    // always SHA-384
    const hashLength = HashObject.SHA_384.length;
    this.digestType = buffer.readInt32BE();
    const {length, bytes} = readLengthAndBytes(buffer.subarray(INT_SIZE), hashLength, hashLength, false);
    this.hash = bytes;

    return INT_SIZE + length;
  }
}

export default HashObject;
