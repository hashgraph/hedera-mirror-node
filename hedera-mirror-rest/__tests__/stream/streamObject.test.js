/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import {INT_SIZE, LONG_SIZE} from '../../stream/constants';
import StreamObject from '../../stream/streamObject';

describe('StreamObject', () => {
  let buffer;
  const classId = BigInt.asIntN(64, BigInt('0xdeadbeaf12345678'));
  const classVersion = 0x12345678;
  const header = Buffer.concat([
    Buffer.from([0x78, 0x56, 0x34, 0x12, 0xaf, 0xbe, 0xad, 0xde]),
    Buffer.from([0x78, 0x56, 0x34, 0x12]),
  ]);

  beforeEach(() => {
    buffer = Buffer.alloc(LONG_SIZE + INT_SIZE);
    buffer.writeBigInt64BE(classId);
    buffer.writeInt32BE(classVersion, LONG_SIZE);
  });

  test('getLength', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.getHeaderLength()).toEqual(LONG_SIZE + INT_SIZE);
  });

  test('getLength', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.getLength()).toEqual(LONG_SIZE + INT_SIZE);
  });

  test('getHeaderLE', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.getHeaderLE()).toEqual(header);
  });

  test('classId', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.classId).toEqual(classId);
  });

  test('classVersion', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.classVersion).toEqual(classVersion);
  });

  test('create from empty buffer', () => {
    expect(() => new StreamObject(Buffer.from([]))).toThrowErrorMatchingSnapshot();
  });

  test('create from truncated buffer', () => {
    expect(() => new StreamObject(Buffer.from([1, 2, 3, 4, 5]))).toThrowErrorMatchingSnapshot();
  });
});
