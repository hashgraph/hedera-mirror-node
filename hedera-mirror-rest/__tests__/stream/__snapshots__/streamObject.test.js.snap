// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamObject create from empty buffer 1`] = `"Attempt to access memory outside buffer bounds"`;

exports[`StreamObject create from truncated buffer 1`] = `"Attempt to access memory outside buffer bounds"`;
